pipe_gaps/pipeline/processes/detect_gaps.py

```diff
@@ -65,33 +65,36 @@ def process_boundaries(
 
         formatted_key = self.boundaries_key().format(key)
 
-        gaps = []
         for messages_pair in boundaries_messages:
-            gaps.extend(self._gd.detect(messages_pair))
+            yield self._gd.detect(messages_pair)
 
-        logger.info(f"Found {len(gaps)} gaps analyzing year boundaries for {formatted_key}...")
+        # logger.info(f"Found {len(gaps)} gaps analyzing year boundaries for {formatted_key}...")
 
         if self._eval_last:
             last_m = max(year_boundaries, key=operator.attrgetter("year")).end
             new_open_gap = self._gd.eval_open_gap(last_m)
 
             if new_open_gap is not None:
-                logger.info(f"Creating 1 open gap for {formatted_key}...")
-                gaps.append(new_open_gap)
-
-        if side_inputs is None:
-            side_inputs = []
-
-        for open_gap in side_inputs:
-            if not key == open_gap["ssvid"]:
-                continue
-
-            logger.info(f"Closing 1 open gap found for {formatted_key}")
-            gaps.append(self._close_open_gap(open_gap, year_boundaries))
-            break
-
-        for gap in gaps:
-            yield gap
+                logger.info(f"Creating 1 new open gap for {formatted_key}...")
+                yield new_open_gap
+
+        side_inputs_list = []
+        if side_inputs is not None:
+            try:
+                side_inputs_list = list(side_inputs[key])
+            except KeyError:
+                # A key was not found for this group.
+                pass
+
+        if len(side_inputs_list) > 0:
+            open_gap = side_inputs_list[0]
+
+            if not isinstance(open_gap, dict):
+                # beam.MultiMap encapsulates value in an iterable of iterables (wtf?).
```

Comment:
`MultiMap` is a multi map: each key may have multiple entries. Beam has no way to know (or ensure, for that matter) than the `PCollection` you transformed into a multi map has one entry per key you computed, so it has to give the broader interface of returning an iterable of records for each key.

BTW, what would be the case where `open_gap` above would be something other than an iterable? Why is that if required?

Reply:
In line `open_gap = side_inputs_list[0]` we are already taking the first element of the iterable. I don´t see why apache beam encapuslates that item inside another iterable. 

We have, for beam, a dict with `key: iterable[iterable]` instead of `key: iterable`.